### PR TITLE
Add support for experimental event versions

### LINKED ISSUE #335: Document how to work with experimental event types
### Description
At the Oct 2022 summit we discussed deployment events and the suggestion was made to start with an experimental v0 version of the proposed EiffelArtifactDeployedEvent (see #322). If we decide to do this there are few practical matters that aren't obvious and therefore should be documented:

* What's the first version of an experimental event? 0.0.1 or 0.1.0 or something else?
* Are we allowed to make any kind of backwards incompatible changes within the v0 line? If so, how should SDKs handle the event versions? For normal non-experimental versions we can use the same data type for all event versions that share a major version because changes will always be forward compatible, but if that no longer is true we can't assume that a particular data type can be used for both, say, 0.1.0 and 0.2.0 of en event.
* ...

### Motivation
If we're going to introduce such a concept we should document _how_ it should be used so that we can be consistent. Documenting things also forces us to really think things through.

### Exemplification
As a maintainer of eiffelevents-sdk-go I'd like to know if and how the experimental events will affect how I generate the SDK, or how the SDK can be used by clients.

### Benefits
Less ambiguity.

### Possible Drawbacks
None.

### PR DESCRIPTION
**Pull request by magnusbaeck**
### Applicable Issues
Fixes #335

### Description of the Change
Adding a section to [versioning.md](https://github.com/eiffel-community/eiffel/blob/master/eiffel-syntax-and-usage/versioning.md) that describes that experimental 0.x.y versions are allowed and how they're supposed to be used. The template file for the documentation Markdown files is updated to include a cautionary banner for 0.x.y. versions. Finally, the YAML schema is updated to include an "experimental" flag for link types that (only) point to experimental events.

I've made an [example commit on the exp-demo branch](https://github.com/magnusbaeck/eiffel/commit/68106cfc5f9e80ba3f4468b990d1b6bf076de34e) so you can see how an experimental event (and links to it) are rendered.

### Alternate Designs
None.

### Benefits
We've seen a need to introduce event types before we're 100% certain how they're supposed to work, and using experimental events for this would avoid calcifying bad designs.

### Possible Drawbacks
This complicates the protocol specification and makes life harder for generated SDKs.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I    have the right to submit it under the open source license    indicated in the file; or

(b) The contribution is based upon previous work that, to the best    of my knowledge, is covered under an appropriate open source    license and I have the right under that license to submit that    work with modifications, whether created in whole or in part    by me, under the same open source license (unless I am    permitted to submit under a different license), as indicated    in the file; or

(c) The contribution was provided directly to me by some other    person who certified (a), (b) or (c) and I have not modified    it.

(d) I understand and agree that this project and the contribution    are public and that a record of the contribution (including all    personal information I submit with it, including my sign-off) is    maintained indefinitely and may be redistributed consistent with    this project or the open source license(s) involved.

Signed-off-by: Magnus Bäck \<magnus.back@axis.com>

### CODE REVIEW

File: event_docs.md.j2 [hunk 1]
```diff
@@ -30,7 +34,7 @@ This section describes which link types are valid for this event type. For detai
 __Required:__ {{ link_data.required | yes_or_no }}  
 __Legal targets:__ {% if link_data.targets.any_type %}Any{% else %}{% for target_type in link_data.targets.types %}{{ target_type | event_link }}{{ ", " if not loop.last else "" }}{% endfor %}{% endif %}  
 __Multiple allowed:__ {{ link_data.multiple | yes_or_no }}  
-__Description:__ {{ link_data.description }}
+__Description:__ {{ link_data.description }}{% if link_data.experimental %} This link type is experimental and may be removed in a future version of the event.{% endif %}
```
Comment (e-backmark-ericsson) [inline event_docs.md.j2:37]: It's actually not the link type that is experimental but it points to an event that is experimental, right?
```suggestion
__Description:__ {{ link_data.description }}{% if link_data.experimental %} This link type targets an experimental event and may be removed in a future version of this event.{% endif %}
```

**Comment by e-backmark-ericsson** [comment]
With the current proposal all source events that link to an experimental event needs to be updated when the experimental event becomes non-experimental. Until now we have never needed to update any source event types just because their target event types change. Is this a behavior we'd like to introduce? I don't have a good alternative solution though.

**Comment by magnusbaeck** [comment]
We'd be updating the documentation of a source event in a way that doesn't affect the schema nor any semantic aspects of the event. It's an helpful hint to the emitter of a source event. I'd consider this similar to typo correction, i.e. something we can change in existing YAML files without issuing any new versions.

**Comment by e-backmark-ericsson** [comment]
> We'd be updating the documentation of a source event in a way that doesn't affect the schema nor any semantic aspects of the event. It's an helpful hint to the emitter of a source event. I'd consider this similar to typo correction, i.e. something we can change in existing YAML files without issuing any new versions.

Fair enough. Agreed.

**Review by e-backmark-ericsson — DISMISSED** [review]

File: event_docs.md.j2 [hunk 2]
```diff
@@ -30,7 +34,7 @@ This section describes which link types are valid for this event type. For detai
 __Required:__ {{ link_data.required | yes_or_no }}  
 __Legal targets:__ {% if link_data.targets.any_type %}Any{% else %}{% for target_type in link_data.targets.types %}{{ target_type | event_link }}{{ ", " if not loop.last else "" }}{% endfor %}{% endif %}  
 __Multiple allowed:__ {{ link_data.multiple | yes_or_no }}  
-__Description:__ {{ link_data.description }}
+__Description:__ {{ link_data.description }}{% if link_data.experimental %} This link type targets an experimental event and may be removed in a future version of this event.{% endif %}
```
Comment (m-linner-ericsson) [inline event_docs.md.j2:37]: Do we want to add something like this?

```suggestion
__Description:__ {{ link_data.description }}{% if link_data.experimental %} This link type targets an experimental event and may be removed in a future version of this event without stepping the major version.{% endif %}
```

or maybe

```suggestion
__Description:__ {{ link_data.description }}{% if link_data.experimental %} This link type targets an experimental event and may be removed in a future version (see [Versioning]}(<correct path>) for details.{% endif %}
```

**Review by m-linner-ericsson — APPROVED** [review]
Was wondering if we want to point out in the link warning text that the link might be removed without stepping the major version. Otherwise we could just point to the longer versioning text for details.

If we don't want to this is approved.

File: event_docs.md.j2 (hunk 2 above)
Comment (magnusbaeck) [reply event_docs.md.j2]: According to the versioning.md update in this PR the link type to the experimental event type may _not_ be removed from the source event type with less than a major version bump. The anything-goes rules only apply to the experimental event type itself.

But linking to versioning.md is totally reasonable so I've pushed a new commit for that. Accidentally force-pushed the wrong commit but I restored the branch and pushed the right commit to the head of the branch.

**Review by e-backmark-ericsson — APPROVED** [review]

**Pull request merged**